Use cstr macro for ffi literal strings

upsuper opened the pull request:
Use `cstr!()` macro with `CStr` to ensure that literal strings used with FFI is properly nul-terminated to avoid cases like #19915.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19925)
<!-- Reviewable:end -->

upsuper committed:
Use cstr macro for ffi literal strings.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: ports/geckolib/lib.rs, ports/geckolib/error_reporter.rs, ports/geckolib/Cargo.toml
 * @emilio: ports/geckolib/lib.rs, ports/geckolib/error_reporter.rs, ports/geckolib/Cargo.toml

upsuper commented:
This currently doesn't pass tidy test because it introduces new version of `syn` and `quote`, so it depends on #19786. I suppose we need to update those crates eventually, given that `cssparser` has been using the new versions.

upsuper commented:
r? @emilio or @nox 

[`cstr`](https://github.com/upsuper/cstr) is a new crate I wrote, so you may want to review code there as well.

nox commented:
There was already a `terminated` crate.

emilio commented:
I think it's nice to use `CStr` in this case, seems easier, we don't need guarantees about encoding...

nox commented:
@emilio Do you mean a macro that returns a CStr, or using CStr directly?

emilio commented:
I meant the macro, but I don't really care. If you think using another crate is a good call, I'm fine with it :)

upsuper commented:
Updated the `cstr` crate to include some more doc and addressed comments from @nox on IRC. The updated `cstr` crate also removes the dependency to `quote` crate, so we have one fewer duplicate version package now. (Although that probably doesn't matter as far as it is still blocked on `syn`.)

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit d95032c has been approved by `emilio`

<!-- @bors-servo r=emilio d95032c9c858456ee9e7706a43c296a986959869 -->

bors-servo commented:
:hourglass: Testing commit d95032c9c858456ee9e7706a43c296a986959869 with merge 9362d2beab8dbec4e52c4581c2cb72a215c78668...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/10916)

CYBAI commented:
```
   Compiling geckoservo v0.0.1 (file:///home/servo/buildbot/slave/linux-dev/build/ports/geckolib)
error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:397:29
    |
397 |             Action::Drop => cstr!("PEDeclDropped").as_ptr(),
    |                             ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:396:29
    |
396 |             Action::Skip => cstr!("PEDeclSkipped").as_ptr(),
    |                             ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:375:27
    |
375 |                     ) => (cstr!("PEColorNotColor"), Action::Nothing),
    |                           ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:368:18
    |
368 |                 (cstr!("PEUnknownAtRule"), Action::Skip),
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:358:18
    |
358 |                 (cstr!("PEDeclDropped"), Action::Nothing),
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:352:25
    |
352 |                         cstr!("PEDeclDropped")
    |                         ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:349:25
    |
349 |                         cstr!("PEMQNoMinMaxWithoutValue")
    |                         ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:346:25
    |
346 |                         cstr!("PEMQExpectedFeatureValue")
    |                         ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:343:25
    |
343 |                         cstr!("PEMQExpectedFeatureName")
    |                         ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:340:25
    |
340 |                         cstr!("PEGatherMediaNotIdent")
    |                         ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:335:33
    |
335 |                 return (prefix, cstr!("PEBadSelectorRSIgnored"), Action::Nothing);
    |                                 ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:329:34
    |
329 |                             Some(cstr!("PENegationBadArg"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:326:34
    |
326 |                             Some(cstr!("PEClassSelNotIdent"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:323:34
    |
323 |                             Some(cstr!("PEPseudoSelBadName"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:320:34
    |
320 |                             Some(cstr!("PEPseudoClassArgNotIdent"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:317:34
    |
317 |                             Some(cstr!("PEPseudoSelEndOrUserActionPC"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:314:34
    |
314 |                             Some(cstr!("PEPseudoSelUnknown"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:311:34
    |
311 |                             Some(cstr!("PESelectorGroupExtraCombinator"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:308:34
    |
308 |                             Some(cstr!("PESelectorGroupNoSelector"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:305:33
    |
305 |                            Some(cstr!("PEUnknownNamespacePrefix"))
    |                                 ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:302:34
    |
302 |                             Some(cstr!("PETypeSelNotType"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:299:34
    |
299 |                             Some(cstr!("PEAttributeNameExpected"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:296:34
    |
296 |                             Some(cstr!("PEAttributeNameOrNamespaceExpected"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:293:34
    |
293 |                             Some(cstr!("PEAttSelBadValue"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:290:34
    |
290 |                             Some(cstr!("PEAttSelNoBar"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:287:34
    |
287 |                             Some(cstr!("PEAttSelUnexpected"))
    |                                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:281:18
    |
281 |                 (cstr!("PEUnknownAtRule"), Action::Nothing)
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:271:18
    |
271 |                 (cstr!("PEAtNSUnexpected"), Action::Nothing)
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:265:18
    |
265 |                 (cstr!("PEBadSelectorKeyframeRuleIgnored"), Action::Nothing),
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:263:18
    |
263 |                 (cstr!("PEKeyframeBadName"), Action::Nothing),
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:261:18
    |
261 |                 (cstr!("PEUnknownFontDesc"), Action::Skip),
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:259:18
    |
259 |                 (cstr!("PEUnknownProperty"), Action::Drop),
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:255:27
    |
255 |                     _ => (cstr!("PEUnknownProperty"), Action::Drop)
    |                           ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:253:26
    |
253 |                         (cstr!("PEValueParsingError"), Action::Drop)
    |                          ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:250:33
    |
250 |                                 cstr!("PEValueParsingError"), Action::Drop)
    |                                 ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:249:38
    |
249 |                         return (Some(cstr!("PEExpectedNoneOrURLOrFilterFunction")),
    |                                      ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:246:33
    |
246 |                                 cstr!("PEValueParsingError"), Action::Drop)
    |                                 ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:245:38
    |
245 |                         return (Some(cstr!("PEColorNotColor")),
    |                                      ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:238:18
    |
238 |                 (cstr!("PEParseDeclarationDeclExpected"), Action::Skip)
    |                  ^^^^

error: cannot find macro `cstr!` in this scope
   --> ports/geckolib/tests/../../../ports/geckolib/error_reporter.rs:382:26
    |
382 |                         (cstr!("PEDeclDropped"), Action::Nothing)
    |                          ^^^^

error: aborting due to 40 previous errors

error: Could not compile `stylo_tests`.
```

upsuper commented:
Fixed the `test-stylo` issue, but it is still blocked on `syn` upgrade.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20022) made this pull request unmergeable. Please resolve the merge conflicts.

upsuper commented:
Looks like #20022 bumpped the versions of the dependencies I need here \o/

upsuper commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 92feec1 has been approved by `emilio`

<!-- @bors-servo r=emilio 92feec1974b6877d7e2688d064b4aa909310ff21 -->

bors-servo commented:
:hourglass: Testing commit 92feec1974b6877d7e2688d064b4aa909310ff21 with merge 5af219e8188f2e657c5711eb5def4c8e8065e29d...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10727), [arm32](http://build.servo.org/builders/arm32/builds/10969), [arm64](http://build.servo.org/builders/arm64/builds/10996), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11027), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7899), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7907), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11096), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3494), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3517), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7759), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7830), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3591), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3535), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6408)
Approved by: emilio
Pushing 5af219e8188f2e657c5711eb5def4c8e8065e29d to master...
